salt/transport/zeromq.py

```diff
@@ -905,12 +942,11 @@ def publish(self, load):
             'Sending payload to publish daemon. jid=%s size=%d',
             load.get('jid', None), len(payload),
         )
-        pub_sock.send(payload)
+        if not self.pub_sock:
+            self.pub_connect()
+        self.pub_sock.send(payload)
```

Comment:
It would be make more sense to send if socket is opened. So I would make `self.pub_connect()` check for socket availability and add logging there too. So then the logic would be:

```python
if self.pub_socket_connect():
    self.pub_sock.send(payload)
```

And if not, you already have an error in the log. You read this better and also one line of code less. :wink: While it is still not pure OOP, yet better.

Reply:
If pub_sock returns true, the socket is opened and connected.

Reply:
Point is, right now it will just crash, if there is a network problem. But if you would put all those timeouts and stuff into generic `socket_connect` you could then better handle exceptions/logging. Like I said, it is up to you.